Source/EphysSocket.cpp

```diff
@@ -162,51 +164,106 @@ bool EphysSocket::startAcquisition()
+    const int total_samples = num_channels * num_samp;
+    const int total_packet_size = total_samples * sizeof(uint16_t);
+    const int packet_ratio = ((total_packet_size + header_size) / MAX_PACKET_SIZE) + 1;
+    const int packet_size = (total_packet_size / packet_ratio) + header_size;
+
+    std::vector<uint16_t> read_buffer;
+    read_buffer.resize(total_samples + 4);
 
-    if (rc == -1)
+    int rc = 0, offset = 0, bytes_sent = 0;
+
+    while (!stop_flag)
     {
-        CoreServices::sendStatusMessage("Ephys Socket: Data shape mismatch");
-        return false;
-    }
-   
-    int k = 0;
-    for (int i = 0; i < num_samp; i++) {
-        for (int j = 0; j < num_channels; j++) {
-            convbuf[k++] = data_scale *  (float)(recvbuf[j*num_samp + i] - data_offset);
+        rc = socket->read(read_buffer.data(), packet_size, true);
+
+        if (rc == -1)
+        {
+            CoreServices::sendStatusMessage("Ephys Socket: Data shape mismatch");
+            error_flag = true;
+            return;
+        }
+
+        offset = read_buffer.at(1) << 16 | read_buffer.at(0);
+        bytes_sent = read_buffer.at(3) << 16 | read_buffer.at(2);
+
+        if (!buffer_flag) recvbuf0.insert(recvbuf0.begin() + (offset / sizeof(uint16_t)), read_buffer.begin() + 4, read_buffer.end()); // This might error if packets are lost
+        else              recvbuf1.insert(recvbuf1.begin() + (offset / sizeof(uint16_t)), read_buffer.begin() + 4, read_buffer.end());
+        
+        if (packet_ratio == 1 || offset + bytes_sent == total_packet_size)
+        {
+            full_flag = true;
+            buffer_flag = !buffer_flag;
         }
-        sampleNumbers.set(i, total_samples++);
-        ttlEventWords.set(i, eventState);
-                
     }
 
-    sourceBuffers[0]->addToBuffer(convbuf, 
-        sampleNumbers.getRawDataPointer(),
-        timestamps.getRawDataPointer(),
-        ttlEventWords.getRawDataPointer(),
-        num_samp, 
-        1);
+    return;
+}
 
-    return true;
+bool EphysSocket::updateBuffer()
+{
+    if (full_flag)
+    {
+        convbuf.clear();
+
+        int k = 0;
+        for (int i = 0; i < num_samp; i++) {
+            for (int j = 0; j < num_channels; j++) {
+                if (buffer_flag)
+                    convbuf.insert(convbuf.begin() + k++, data_scale * (float)(recvbuf0.at(j * num_samp + i) - data_offset));
+                else
+                    convbuf.insert(convbuf.begin() + k++, data_scale * (float)(recvbuf1.at(j * num_samp + i) - data_offset));
+            }
+            sampleNumbers.set(i, total_samples++);
+            ttlEventWords.set(i, eventState);
+        }
+
+        sourceBuffers[0]->addToBuffer(convbuf.data(),
+            sampleNumbers.getRawDataPointer(),
+            timestamps.getRawDataPointer(),
+            ttlEventWords.getRawDataPointer(),
+            num_samp,
+            1);
+
+        full_flag = false;
+
+        if (buffer_flag)  recvbuf0.clear();
```

Comment:
Why are you clearing the buffer? It will be overwritten on the next pass. No need.